File: 프로그래머스/lv2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.py
from collections import deque
graph = {}

def make_graph(wire, n):
    global graph
    
    for i in range(1, n+1):
        graph[i] = []
    
    for w in wire:
        graph[w[0]].append(w[1])
        graph[w[1]].append(w[0])

def bfs(start, graph, visited):
    queue = deque([start])
    visited[start] = True
    # print(visited)
    while queue:
        v = queue.popleft()
        
        for i in graph[v]:
            if visited[i] == False:
                queue.append(i)
                visited[i] = True

def solution(n, wires):
    global graph
    answer = 100
    wires = sorted(wires, key = lambda x : (x[0],x[1]))
    
    for i in range(len(wires)):
        wire = wires[:i] + wires[i+1:]
        make_graph(wire, n)
        print(graph[1])
        node = set([])
        for j in range(1, n+1):
            visited = [False] * (n+1)
            bfs(j, graph, visited)
            node.add(visited.count(True))
            
        if max(node) - min(node) < answer:
            answer = max(node) - min(node)
    
    return answer

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.js
function solution(n, wires) {
    let answer = Infinity
    
    const graph = Array(n+1).fill().map(() => [])

    for (const [v1, v2] of wires){
        graph[v1].push(v2)
        graph[v2].push(v1)
    }
    
    function dfs(graph, start, excluded) {
        let visited = new Set()
        let queue = [start]
        visited.add(start)
        
        let count = 0
        while(queue.length > 0){
            let node = queue.shift()
            count += 1
            
            for (const neighbor of graph[node]) {
                if ((node == excluded[0] && neighbor == excluded[1]) || 
                   (node == excluded[1] && neighbor == excluded[0]) ) {
                    continue
                }
                
                if (!visited.has(neighbor)) {
                    visited.add(neighbor)
                    queue.push(neighbor)
                }
            }
        }
        
        return count
    }
    
    for (const [v1, v2] of wires) {
        const groupSize = dfs(graph, 1, [v1, v2])
        const otherGroupSize = n - groupSize
        const diff = Math.abs(groupSize - otherGroupSize)

        answer = Math.min(answer, diff)
        
    }
    
    return answer
}